core/parentchain/block-importer/src/parentchain_block_importer.rs

```diff
@@ -0,0 +1,230 @@
+	type ParentchainBlockType: BlockT;
+
+	/// Import parentchain blocks to the light-client (validator):
+	/// * iterates over parentchain blocks and scans for relevant extrinsics
+	/// * validates and execute those extrinsics (containing indirect calls), mutating state
+	/// * sends `PROCESSED_PARENTCHAIN_BLOCK` extrinsics that include the merkle root of all processed calls
+	fn import_parentchain_blocks(
+		&self,
+		blocks_to_import: Vec<SignedBlockG<Self::ParentchainBlockType>>,
+	) -> Result<()>;
+}
+
+/// Parentchain block import implementation.
+pub struct ParentchainBlockImporter<
+	PB,
+	ValidatorAccessor,
+	OCallApi,
+	StfExecutor,
+	ExtrinsicsFactory,
+	IndirectCallsExecutor,
+> where
+	PB: BlockT<Hash = H256>,
+	NumberFor<PB>: BlockNumberOps,
+	ValidatorAccessor: ValidatorAccess<PB>,
+	OCallApi: EnclaveOnChainOCallApi + EnclaveAttestationOCallApi,
+	StfExecutor: StfUpdateState + StfExecuteTrustedCall + StfExecuteShieldFunds,
+	ExtrinsicsFactory: CreateExtrinsics,
+	IndirectCallsExecutor: ExecuteIndirectCalls,
+{
+	validator_accessor: Arc<ValidatorAccessor>,
+	ocall_api: Arc<OCallApi>,
+	stf_executor: Arc<StfExecutor>,
+	extrinsics_factory: Arc<ExtrinsicsFactory>,
+	indirect_calls_executor: Arc<IndirectCallsExecutor>,
+	_phantom: PhantomData<PB>,
+}
+
+impl<PB, ValidatorAccessor, OCallApi, StfExecutor, ExtrinsicsFactory, IndirectCallsExecutor>
+	ParentchainBlockImporter<
+		PB,
+		ValidatorAccessor,
+		OCallApi,
+		StfExecutor,
+		ExtrinsicsFactory,
+		IndirectCallsExecutor,
+	> where
+	PB: BlockT<Hash = H256, Header = ParentchainHeader>,
+	NumberFor<PB>: BlockNumberOps,
+	ValidatorAccessor: ValidatorAccess<PB>,
+	OCallApi: EnclaveOnChainOCallApi + EnclaveAttestationOCallApi,
+	StfExecutor: StfUpdateState + StfExecuteTrustedCall + StfExecuteShieldFunds,
+	ExtrinsicsFactory: CreateExtrinsics,
+	IndirectCallsExecutor: ExecuteIndirectCalls,
+{
+	pub fn new(
+		validator_accessor: Arc<ValidatorAccessor>,
+		ocall_api: Arc<OCallApi>,
+		stf_executor: Arc<StfExecutor>,
+		extrinsics_factory: Arc<ExtrinsicsFactory>,
+		indirect_calls_executor: Arc<IndirectCallsExecutor>,
+	) -> Self {
+		ParentchainBlockImporter {
+			validator_accessor,
+			ocall_api,
+			stf_executor,
+			extrinsics_factory,
+			indirect_calls_executor,
+			_phantom: Default::default(),
+		}
+	}
+}
+
+impl<PB, ValidatorAccessor, OCallApi, StfExecutor, ExtrinsicsFactory, IndirectCallsExecutor>
+	ImportParentchainBlocks
+	for ParentchainBlockImporter<
+		PB,
+		ValidatorAccessor,
+		OCallApi,
+		StfExecutor,
+		ExtrinsicsFactory,
+		IndirectCallsExecutor,
+	> where
+	PB: BlockT<Hash = H256, Header = ParentchainHeader>,
+	NumberFor<PB>: BlockNumberOps,
+	ValidatorAccessor: ValidatorAccess<PB>,
+	OCallApi: EnclaveOnChainOCallApi + EnclaveAttestationOCallApi,
+	StfExecutor: StfUpdateState + StfExecuteTrustedCall + StfExecuteShieldFunds,
+	ExtrinsicsFactory: CreateExtrinsics,
+	IndirectCallsExecutor: ExecuteIndirectCalls,
+{
+	type ParentchainBlockType = PB;
+
```

Comment:
```suggestion
/// Imports parentchain blocks, by iterating through them and their extrinsics. Upon finding a specified extrinsic, it will be executed on the state with the StfExecutor.  The following steps are executed:
///  For every block: 
      - check merkle proof of the header
      - submit header into the light client
      - update parentchain state of the stf executor
      - create a `confirm_processed_parentchain_block` extrinsic and send it to the parentchain node
/// For every extrinsic:
      - check if it has been included previosuly in the validator cache
      - if specified, execute with the stf executor
```

Reply:
with everything commented 😋

Reply:
hmm to be honest, I'm not a big fan of replicating the code in the comments. For every reader it should be clear that we iterate over all blocks. And that we use the `StfExecutor` is an implementation detail, not interesting from a caller perspective. Interesting are things like pre-conditions, post-conditions, or in general information that you cannot already get from the signature.

Reply:
So I wouldn't even include that list of every step that the function does internally. Only what it does from a caller perspective.

Reply:
Ok, I can agree with the part 

> The code should be written well enough that these steps can be deduced easily.

It's just that this part of the code will most likely be one of the most important one for developers using our worker. So comments and code should really be clear.

Reply:
That Blocks are iterated through may be clear, but not the extrinsic part. So if you include somewhere that "import_block" also means that specified (!) extrsinics will be executed, I'd be delighted.

Reply:
The trait where this function is defined, has that information now:
```
	/// Import parentchain blocks to the light-client (validator):
	/// * Scans the block for extrinsics containing indirect calls
	/// * Validates and execute those extrinsics, mutating state
	/// * Sends `PROCESSED_PARENTCHAIN_BLOCK` extrinsics that include the merkle root of all processed calls
	fn import_parentchain_blocks(
		&self,
		blocks_to_import: Vec<SignedBlockG<Self::ParentchainBlockType>>,
	) -> Result<()>;
```

So maybe the question is more, do we document the trait function, or the implementation function? In other languages, you have a special tag in the documentation, saying 'just take the documentation from the interface/trait that I'm implementing here'. Because maintaining a single documentation block is already challenging at times, but if you have everything duplicated to each implementation, it becomes unmaintainable.

Reply:
Trait function I'd say. That's also what the IDE (atleast VS code) shows as function description.

Reply:
I would say trait function, this also implies that we leave out the implementation details, which, I think, should nod be there.

Reply:
Could you change 
`/// * Scans the block for extrinsics containing indirect calls`

To a more general
`/// * Scans the blocks for relevant extrinsics` ? 
Because what extrinsic will be scanned for, is very use case specific.

And add 
`/// * Validates and includes block headers into the light client` . I think that's a very basic functionality.